File: spring-web/src/main/java/com/example/springwebex/filter/ExceptionResolverFilter.java
package com.example.springwebex.filter;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

import org.springframework.stereotype.Component;

import com.example.springwebex.exception.ecode.ErrorCode;
import com.example.springwebex.model.restresp.ResponseJsonDto;

import jakarta.servlet.Filter;
import jakarta.servlet.FilterChain;
import jakarta.servlet.FilterConfig;
import jakarta.servlet.ServletException;
import jakarta.servlet.ServletRequest;
import jakarta.servlet.ServletResponse;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Component
public class ExceptionResolverFilter implements Filter {
    private static final String INTERNAL_ERROR_RESPONSE;

    static {
        ResponseJsonDto<?> responseObject = ResponseJsonDto.from(
                String.valueOf(ErrorCode.UNKNOWN.getCode()),
                ErrorCode.UNKNOWN.getDefaultMessage());

        try {
            INTERNAL_ERROR_RESPONSE = responseObject.toString();
        } catch (Exception e) {
            log.error("failed to initalize class member.", e);
            throw new ExceptionInInitializerError(e);
        }
    }

    @Override
    public void init(FilterConfig filterConfig) throws ServletException {
        // do nothing
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
            throws IOException, ServletException {

        try {
            chain.doFilter(request, response);
        } catch (Exception e) {
            log.error(e.getMessage());
            response.getOutputStream()
                    .write(INTERNAL_ERROR_RESPONSE.getBytes(StandardCharsets.UTF_8));
        }
    }

    @Override
    public void destroy() {
        // do nothing
    }

}

File: spring-web-kotlin/src/main/kotlin/com/example/springwebex/filter/ExceptionResolverFilter.kt
package com.example.springwebex.filter

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Component
import jakarta.servlet.Filter
import jakarta.servlet.FilterChain
import jakarta.servlet.FilterConfig
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import jakarta.servlet.http.HttpServletResponse
import io.github.oshai.kotlinlogging.KotlinLogging

@Component
class ExceptionResolverFilter : Filter {

    private val log = KotlinLogging.logger {}
    private val objectMapper = ObjectMapper()

    override fun init(filterConfig: FilterConfig?) {
        log.debug { "ExceptionResolverFilter initialized" }
    }

    override fun doFilter(
        request: ServletRequest?,
        response: ServletResponse?,
        chain: FilterChain?
    ) {
        try {
            chain?.doFilter(request, response)
        } catch (e: Exception) {
            log.error(e) { "Exception occurred in filter" }
            val httpResponse = response as? HttpServletResponse
            if (httpResponse != null) {
                httpResponse.status = HttpServletResponse.SC_INTERNAL_SERVER_ERROR
                httpResponse.contentType = "application/json"
                try {
                    val errorResponse = mapOf(
                        "error" to "Internal Server Error",
                        "message" to (e.message ?: "An error occurred"),
                        "timestamp" to System.currentTimeMillis()
                    )
                    httpResponse.writer.write(objectMapper.writeValueAsString(errorResponse))
                } catch (ioe: Exception) {
                    log.error(ioe) { "Error writing response" }
                }
            }
        }
    }

    override fun destroy() {
        log.debug { "ExceptionResolverFilter destroyed" }
    }
}
